File: spring-data/spring-data-jpa/jpa-basic-sample/src/main/java/org/example/shop/controller/MemberController.java
package org.example.shop.controller;

import java.util.List;
import javax.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.example.shop.domain.Address;
import org.example.shop.domain.Member;
import org.example.shop.service.MemberService;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;

@Slf4j
@Controller
@RequiredArgsConstructor
public class MemberController {

  private final MemberService memberService;

  @GetMapping("/members/new")
  public String createForm(Model model) {
    model.addAttribute("memberForm", new MemberForm());
    return "members/createMemberForm";
  }

  @PostMapping("/members/new")
  public String create(@Valid MemberForm form, BindingResult bindingResult) {

    if (bindingResult.hasErrors()) {
      log.info("Error = {}", bindingResult);
      return "members/createMemberForm";
    }

    Address address = new Address(form.getCity(), form.getStreet(), form.getZipcode());

    Member member = Member.builder()
      .name(form.getName())
      .address(address)
      .build();

    memberService.join(member);
    return "redirect:/";
  }

  @GetMapping("/members")
  public String list(Model model) {
    List<Member> members = memberService.findMembers();
    model.addAttribute("members", members);
    return "members/memberList";
  }
}

File: spring-mvc/spring-mvc-2/src/main/kotlin/com/springthymeleaf/login/web/member/MemberController.kt
package com.springthymeleaf.login.web.member

import com.springthymeleaf.login.domain.member.Member
import com.springthymeleaf.login.domain.member.MemberRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Controller
import org.springframework.validation.BindingResult
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping

@Controller
@RequestMapping("/members")
class MemberController(
  private val memberRepository: MemberRepository
) {

  private val log = LoggerFactory.getLogger(javaClass)

  @GetMapping("/add")
  fun addForm(@ModelAttribute("member") member: Member): String {
    return "members/add-member-form"
  }

  @PostMapping("/add")
  fun add(@Validated @ModelAttribute member: Member, bindingResult: BindingResult): String {
    if (bindingResult.hasErrors())
      return "members/add-member-form"

    log.info("request member = {}", member)

    memberRepository.save(member)
    return "redirect:/"
  }
}